modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java

```diff
@@ -4411,6 +4486,55 @@ private void nodeCheckError(TcpDiscoveryNode node, String errMsg, String sndMsg)
             }
         }
 
+        /** */
+        private void trySendMessageDirectlyToAddrs(
+            Collection<InetSocketAddress> addrs,
+            @Nullable TcpDiscoveryNode node,
+            TcpDiscoveryAbstractMessage msg
+        ) {
+            IgniteSpiException ex = null;
+
+            for (InetSocketAddress addr : addrs) {
+                try {
+                    IgniteSpiOperationTimeoutHelper timeoutHelper = new IgniteSpiOperationTimeoutHelper(spi, true);
+
+                    sendMessageDirectly(msg, addr, timeoutHelper);
+
+                    if (node != null)
+                        node.lastSuccessfulAddress(addr);
+
+                    return;
+                }
+                catch (IgniteSpiException e) {
+                    ex = e;
+                }
+            }
+
+            if (ex != null)
+                throw ex;
+        }
+
+        /**
+         * Tries to send a message to all node's available addresses.
+         *
+         * @param nodeId Node ID to send message to.
+         * @param addrs Node addresses, used when node could not be found in <code>ring</code>.
+         * @param msg Message.
+         * @throws IgniteSpiException Last failure if all attempts failed.
+         */
+        private void trySendMessageDirectly(UUID nodeId, Collection<InetSocketAddress> addrs, TcpDiscoveryAbstractMessage msg) {
```

Comment:
First: need to refactor arguments' list as it doesn't fit into one line.

Second: I don't like that first argument of this method is addrs collection when method `trySendMessageDirectlyToAddrs` has the same argument at the first position. So please rearrange arguments list to make it easier to match them visually.